pyVHDLParser/Blocks/Reference/Context.py

```diff
@@ -27,139 +27,329 @@
 
 	@classmethod
-	def stateWhitespace1(cls, parserState: TokenToBlockParser):
+	def stateContextName(cls, parserState: TokenToBlockParser):
 		token = parserState.Token
+
+		if _isNonCodeToken(token):
+			return
+
+		if isinstance(token, (WordToken, ExtendedIdentifier)):
+			parserState.NextState =     cls.stateWhitespaceOrSemicolonOrDot
+			return
+
+		raise BlockParserException("Expected context name (identifier).", token)
+
+	@classmethod
+	def stateWhitespaceOrSemicolonOrDot(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if _isNonCodeToken(token):
+			parserState.NextState =        cls.stateIsOrSemicolonOrDot
+			return
+
+		if isinstance(token, CharacterToken) and token == ";":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			parserState.Pop()
+			return
+		elif isinstance(token, CharacterToken) and token == ".":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			return
+
+		raise BlockParserException("Expected whitespace after context name (identifier).", token)
+
+	@classmethod
+	def stateIsOrSemicolonOrDot(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if parserState.HandleNonCodeTokens(cls):
+			return
+
+		if isinstance(token, WordToken) and (token == "is"):
+			parserState.ReparseFromTokenMarker(DeclarationStartBlock.stateFromStartBlock)
+			return
+		elif isinstance(token, CharacterToken) and token == ";":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			parserState.Pop()
+			return
+		elif isinstance(token, CharacterToken) and token == ".":
+			parserState.ReparseFromTokenMarker(ReferenceStartBlock.stateFromStartBlock)
+			return
+
+		raise BlockParserException("Expected context name (identifier).", token)
+
+
+#
+# Context references
+#
+@export
+class ReferenceStartBlock(Block):
+	@classmethod
+	def stateFromStartBlock(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if isinstance(token, ContextKeyword):
+			parserState.NextState =   cls.stateWhitespace
+			return
+
+		# The 'context' keyword was already detected. If it is not there anymore, we found a bug.
+		assert False, "Expected keyword CONTEXT."
+
+	@classmethod
+	def stateContextKeyword(cls, parserState: TokenToBlockParser):
+		cls.stateWhitespace(parserState)
+
+	@classmethod
+	def stateWhitespace(cls, parserState: TokenToBlockParser):
+		if parserState.HandleNonCodeTokens(cls, multiPart=False):
+			parserState.NextState =   ReferenceNameBlock.stateLibOrContextName
+			return
+
+		# This condition is also guaranteed. Otherwise `StartBlock.stateContextKeyword` would have thrown an error.
+		assert False, "Expected whitespace after keyword CONTEXT."
+
+
+@export
+class ReferenceNameBlock(Block):
+	KEYWORDS = None
+
+	@classmethod
+	def stateLibOrContextName(cls, parserState: TokenToBlockParser):
+		token = parserState.Token
+
+		if parserState.HandleNonCodeTokens(None):

```

Comment:
That's my take on handling non-code tokens, without repeating the same 20 lines in every other state. If a block type is passed as the first parameter, an (multi part) instance of that type is created before any whitespace/comment block.